File: 206_reverse_linked_list/reverse_linked_list.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def reverse_linked_list(self, head: ListNode) -> ListNode:
        prev = None
        current = head
        while current:
            nxt = current.next
            current.next = prev
            prev = current
            current = nxt
        return prev


if __name__ == "__main__":
    head = ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5)))))
    solution = Solution()
    reversed_head = solution.reverse_linked_list(head)

    result = []
    while reversed_head:
        result.append(reversed_head.val)
        reversed_head = reversed_head.next
    print(result)

File: 206_reverse_linked_list/reverse_linked_list.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

type Solution struct{}

func (s *Solution) reverseLinkedList(head *ListNode) *ListNode {
	var prev *ListNode
	current := head
	for current != nil {
		next := current.Next
		current.Next = prev
		prev = current
		current = next
	}
	return prev
}

func main() {
	head := &ListNode{1, &ListNode{2, &ListNode{3, &ListNode{4, &ListNode{5, nil}}}}}
	s := Solution{}
	reversed := s.reverseLinkedList(head)

	var result []int
	for reversed != nil {
		result = append(result, reversed.Val)
		reversed = reversed.Next
	}
	fmt.Println(result)
}
